pulsar-common/src/main/java/org/apache/pulsar/common/naming/TopicName.java

```diff
@@ -231,7 +231,7 @@ public String getEncodedLocalName() {
     }
 
     public TopicName getPartition(int index) {
-        if (index == -1 || this.toString().contains(PARTITIONED_TOPIC_SUFFIX)) {
+        if (index == -1 || this.toString().endsWith(PARTITIONED_TOPIC_SUFFIX + index)) {
```

Comment:
> we will not automatically create a partitioned topic for the DLQ topic even if users are enabled partitioned topic auto-creation.

Yes, [PIP-263: Just auto-create no-partitioned DLQ And Prevent auto-create a DLQ for a DLQ]( https://github.com/apache/pulsar/issues/20033) try to do this.

Reply:
> What I thought was the PR intent to allow the partitioned name can contain -partition- keyword to resolve the issue. But it will introduce a new unreasonable rule to the partitioned topic.

I think there are three things:
- Our rule for a partition name and how to check which index it is: end with `-partition-x` or contain `-partition-x`
  - The current logic is incorrect, and current PR can fix it. 
- Whether to allow to create a topic whose name contains `-paritition-`
  - the behavior is denied for Admin API and is allowed for client(such as create by a cmd-subscribe), you can see the tests `testInfiniteHttpCallGetSubscriptions2`, `testInfiniteHttpCallGetSubscriptions3` in current PR
- Whether only create non-partitioned topics for the Dead Letter or Retry Letter
  - [PIP-263: Just auto-create no-partitioned DLQ And Prevent auto-create a DLQ for a DLQ]( https://github.com/apache/pulsar/issues/20033) try to do this.